samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java

```diff
@@ -371,47 +389,54 @@ private boolean hasActiveContainerPlacementAction(String processorId) {
     return Optional.ofNullable(this.actions.get(processorId));
   }
 
+  private void updateContainerPlacementActionStatus(ContainerPlacementMetadata metadata,
+      ContainerPlacementMessage.StatusCode statusCode, String responseMessage) {
+    metadata.setActionStatus(statusCode, responseMessage);
+    writeContainerPlacementResponseMessage(metadata.getRequestMessage(), statusCode, responseMessage);
+    LOG.info("Status updated for ContainerPlacement action: {}", metadata);
+  }
+
+  private void writeContainerPlacementResponseMessage(ContainerPlacementRequestMessage requestMessage,
+      ContainerPlacementMessage.StatusCode statusCode, String responseMessage) {
+    containerPlacementMetadataStore.writeContainerPlacementResponseMessage(
+        ContainerPlacementResponseMessage.fromContainerPlacementRequestMessage(requestMessage, statusCode,
+            responseMessage, System.currentTimeMillis()));
+  }
+
   /**
    * A valid container placement action is only issued for a running processor with a valid processor id which has no
-   * in flight container requests. Duplicate actions are handled by deduping on uuid
-   *
-   * TODO: SAMZA-2402: Disallow pending Container Placement actions in metastore on job restarts
+   * in flight container requests. Duplicate actions are handled by deduping on uuid. If there is an existing inflight
+   * request or container is pending a start, the container placement request is queued to be executed in future.
    */
-  private Pair<ContainerPlacementMessage.StatusCode, String> checkValidControlAction(String processorId, String destinationHost, UUID uuid) {
-    String errorMessagePrefix =
-        String.format("ControlAction to move or restart container with processor id %s to host %s is rejected due to",
-            processorId, destinationHost);
+  private Optional<Pair<ContainerPlacementMessage.StatusCode, String>> validatePlacementAction(ContainerPlacementRequestMessage requestMessage) {
+    String errorMessagePrefix = String.format("ContainerPlacement request: %s is rejected due to", requestMessage);
```

Comment:
For a method that is called validatePlacementAction a return value of null/empty means the action is valid?
seems like weird semantics.

Reply:
forgot to push a commit that refactors this method to two methods: on that dequeues and other that validates, please have a look